### PR TITLE
rustdoc: properly detect input file exists

### LINKED ISSUE #27014: Rustdoc error output is unnecessarily verbose when encountering unknown file
What `rustc` displays when encountering an unknown file:

```
/tmp $ rustc filename-that-does-not-correspond-to-a-file
error: couldn't read "filename-that-does-not-correspond-to-a-file": No such file or directory (os error 2)
```

...and what `rustdoc` displays:

```
/tmp $ rustdoc filename-that-does-not-correspond-to-a-file
error: couldn't read "filename-that-does-not-correspond-to-a-file": No such file or directory (os error 2)
thread '<unnamed>' panicked at 'Box<Any>', ../src/libsyntax/diagnostic.rs:211
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "rustc failed"', ../src/libcore/result.rs:731
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Any', ../src/libcore/result.rs:731
```

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #27014 

r? @alexcrichton 

I'm not 100% sure there's not a better way to do this, but it works.

Also, I wasn't sure how, where, or if to write a test for this.

### CODE REVIEW

**Comment by alexcrichton** [comment]
The compiler's actually already spitting out the right error message here, it's just that none of the subsequent panic messages are being suppressed. To solve this I think it'd be better to handle the panic messages in the same manner as the compiler, only emitting some of them for major errors.

**Comment by steveklabnik** [comment]
@alexcrichton what about this second commit instead? It now prints out

```
$ x86_64-unknown-linux-gnu/stage1/bin/rustdoc fake
No such file or directory (os error 2)
```

Handling bad input proactively just seems like a good idea to me, personally, rather than letting some inner thing panic.

**Comment by alexcrichton** [comment]
rustdoc currently has a bad error message for *any* panic on behalf of the compiler (e.g. any compiler error message), so I think it'd be best to fix both issues instead of just special casing this one

**Comment by steveklabnik** [comment]
word. i'll try to figure out where that is.

**Comment by alexcrichton** [comment]
All of rustc's logic is in the [`monitor` function](https://github.com/rust-lang/rust/blob/master/src/librustc_driver/lib.rs#L781-L843), so this may be as simple as just calling that when rustdoc spawns a thread.

**Comment by steveklabnik** [comment]
Nice! using `monitor` now makes it look like this:

```
$ x86_64-unknown-linux-gnu/stage1/bin/rustdoc fake
error: couldn't read "fake": No such file or directory (os error 2)
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Any', src/libcore/result.rs:731
```

That last extra message happens because of https://github.com/rust-lang/rust/blob/master/src/librustdoc/lib.rs#L135

Given that we're setting up a 32MB stack here, I guess I can't turn this spawn into a monitor either, eh? What should I do here?

**Comment by alexcrichton** [comment]
I think it's fine to change that `unwrap` to just a `unwrap_or(101)`

**Comment by steveklabnik** [comment]
ahhh yes. I just started `make check` on the rollup, but i'll do that after. anything else I should take care of, assuming that works? I'll squash and such too

**Comment by alexcrichton** [comment]
Nah I think that should do it!

File: src/librustdoc/lib.rs [hunk 1]
```diff
@@ -56,7 +56,7 @@ extern crate serialize as rustc_serialize; // used by deriving
 use std::cell::RefCell;
 use std::collections::HashMap;
 use std::env;
-use std::fs::File;
+use std::fs::{self, File};
```
Comment (tamird) [inline src/librustdoc/lib.rs]: doesn't look like this change is needed

Comment (steveklabnik) [inline src/librustdoc/lib.rs]: yes, it's since been removed

**Comment by steveklabnik** [comment]
@alexcrichton r?

**Pull request merged**